File: bumperbot_ws/src/bumperbot_controller/src/noisy_controller.cpp
#include "bumperbot_controller/noisy_controller.hpp"
#include <random>
#include <memory>

NoisyController::NoisyController(const std::string &name) : Node(name)
{
    // Declare parameters
    declare_parameter("wheel_radius", 0.033);
    declare_parameter("wheel_separation", 0.17);

    // Get parameters
    wheel_radius_ = get_parameter("wheel_radius").as_double();
    wheel_separation_ = get_parameter("wheel_separation").as_double();

    RCLCPP_INFO_STREAM(get_logger(), "Using wheel radius " << wheel_radius_);
    RCLCPP_INFO_STREAM(get_logger(), "Using wheel separation " << wheel_separation_);

    // Initialize publisher and subscriber
    joint_sub_ = create_subscription<sensor_msgs::msg::JointState>(
        "/joint_states", 10,
        std::bind(&NoisyController::jointCallback, this, std::placeholders::_1));

    odom_pub_ = create_publisher<nav_msgs::msg::Odometry>("/bumperbot_controller/odom_noisy", 10);

    left_wheel_prev_pos_ = 0;
    right_wheel_prev_pos_ = 0;
    prev_time_ = get_clock()->now();
    x_ = 0.0;
    y_ = 0.0;
    theta_ = 0.0;

    q.setRPY(0, 0, theta_);

    odom_msgs_.header.frame_id = "odom";
    odom_msgs_.child_frame_id = "base_footprint_noisy";
    odom_msgs_.pose.pose.orientation.x = 0.0;
    odom_msgs_.pose.pose.orientation.y = 0.0;
    odom_msgs_.pose.pose.orientation.z = 0.0;
    odom_msgs_.pose.pose.orientation.w = 1.0;

    br_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
    transform_stamped_.header.frame_id = "odom";
    transform_stamped_.child_frame_id = "base_footprint_noisy";
}

void NoisyController::jointCallback(const sensor_msgs::msg::JointState &msg)
{
    std::default_random_engine noise_generator(std::chrono::system_clock::now().time_since_epoch().count());
    std::normal_distribution<double> left_encoder_noise(0.0, 0.005);
    std::normal_distribution<double> right_encoder_noise(0.0, 0.005);

    double wheel_encoder_right = msg.position.at(0) + right_encoder_noise(noise_generator);
    double wheel_encoder_left = msg.position.at(1) + left_encoder_noise(noise_generator);

    double dp_right = wheel_encoder_right - right_wheel_prev_pos_;
    double dp_left = wheel_encoder_left - left_wheel_prev_pos_;

    rclcpp::Time msg_time = msg.header.stamp;
    rclcpp::Duration dt = msg_time - prev_time_;

    right_wheel_prev_pos_ = wheel_encoder_right;
    left_wheel_prev_pos_ = wheel_encoder_left;
    prev_time_ = msg_time;

    double fi_left = dp_left / dt.seconds();
    double fi_right = dp_right / dt.seconds();

    double V = wheel_radius_ / 2 * (fi_right + fi_left);
    double w = wheel_radius_ / wheel_separation_ * (fi_right - fi_left);

    double d_s = wheel_radius_ / 2 * (dp_right + dp_left);
    double d_theta = wheel_radius_ / wheel_separation_ * (dp_right - dp_left);
    theta_ += d_theta;
    x_ += d_s * std::cos(theta_);
    y_ += d_s * std::sin(theta_);

    q.setRPY(0, 0, theta_);
    odom_msgs_.pose.pose.orientation.x = q.x();
    odom_msgs_.pose.pose.orientation.y = q.y();
    odom_msgs_.pose.pose.orientation.z = q.z();
    odom_msgs_.pose.pose.orientation.w = q.w();
    odom_msgs_.pose.pose.position.x = x_;
    odom_msgs_.pose.pose.position.y = y_;
    odom_msgs_.header.stamp = get_clock()->now();
    odom_msgs_.twist.twist.linear.x = V;
    odom_msgs_.twist.twist.angular.z = w;

    odom_pub_->publish(odom_msgs_);

    transform_stamped_.transform.translation.x = x_;
    transform_stamped_.transform.translation.y = y_;
    transform_stamped_.transform.rotation.x = q.x();
    transform_stamped_.transform.rotation.y = q.y();
    transform_stamped_.transform.rotation.z = q.z();
    transform_stamped_.transform.rotation.w = q.w();
    transform_stamped_.header.stamp = get_clock()->now();
    br_->sendTransform(transform_stamped_);
}

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NoisyController>("noisy_controller");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: bumperbot_ws/src/bumperbot_controller/bumperbot_controller/noisy_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import  TransformStamped
from sensor_msgs.msg import JointState
import numpy as np
from rclpy.time import Time
from rclpy.constants import S_TO_NS
import math
from nav_msgs.msg import Odometry
from tf_transformations import quaternion_from_euler
from tf2_ros import TransformBroadcaster


class NoisyController(Node):
    def __init__(self):
        # Ensure the node is initialized with the correct name
        super().__init__("noisy_controller")

        self.declare_parameter("wheel_radius", 0.033)
        self.declare_parameter("wheel_separation", 0.17)

        self.wheel_radius_ = self.get_parameter("wheel_radius").get_parameter_value().double_value
        self.wheel_separation_ = self.get_parameter("wheel_separation").get_parameter_value().double_value

        self.get_logger().info("Using wheel radius %f" % self.wheel_radius_)
        self.get_logger().info("Using wheel separation %f" % self.wheel_separation_)

        self.left_wheel_prev_pos_ = 0
        self.right_wheel_prev_pos_ = 0
        self.prev_time_ = self.get_clock().now()

        self.x_ = 0.0
        self.y_ = 0.0
        self.theta_ = 0


        self.joint_sub_ = self.create_subscription(JointState, "joint_states", self.jointCallback, 10)
        self.odom_pub_ = self.create_publisher(Odometry, "bumperbot_controller/odom_noisy", 10)

        self.odom_msgs_ = Odometry()
        self.odom_msgs_.header.frame_id = "odom"
        self.odom_msgs_.child_frame_id = "base_footprint_ekf"
        self.odom_msgs_.pose.pose.orientation.x = 0.0
        self.odom_msgs_.pose.pose.orientation.y = 0.0
        self.odom_msgs_.pose.pose.orientation.z = 0.0
        self.odom_msgs_.pose.pose.orientation.w = 1.0

        self.br_ = TransformBroadcaster(self)
        self.transform_stamped_= TransformStamped()
        self.transform_stamped_.header.frame_id = "odom"
        self.transform_stamped_.child_frame_id = "base_footprint_noisy"



    def jointCallback(self, msg):
        wheel_encoder_right = msg.position[0]+np.random.normal(0,0.005)
        wheel_encoder_left = msg.position[1]+np.random.normal(0,0.005)

        dp_right= wheel_encoder_right - self.right_wheel_prev_pos_
        dp_left = wheel_encoder_left - self.left_wheel_prev_pos_
        dt = Time.from_msg(msg.header.stamp) - self.prev_time_

        # self.right_wheel_prev_pos_ = msg.position[0]
        # self.left_wheel_prev_pos_ = msg.position[1]

        self.right_wheel_prev_pos_ = wheel_encoder_right
        self.left_wheel_prev_pos_ = wheel_encoder_left
        
        self.prev_time_= Time.from_msg(msg.header.stamp)

        fi_left = dp_left/(dt.nanoseconds/S_TO_NS)
        fi_right = dp_right/(dt.nanoseconds/S_TO_NS)

        V = self.wheel_radius_/2*fi_right + self.wheel_radius_/2*fi_left
        w = self.wheel_radius_/self.wheel_separation_*fi_right - self.wheel_radius_/self.wheel_separation_*fi_left

        d_s = self.wheel_radius_/2*dp_right +self.wheel_radius_/2*dp_left
        d_theta = self.wheel_radius_/self.wheel_separation_*(dp_right-dp_left)
        self.theta_ +=d_theta
        self.x_+=d_s*math.cos(self.theta_)
        self.y_+=d_s*math.sin(self.theta_)


        q = quaternion_from_euler(0,0,self.theta_)
        self.odom_msgs_.pose.pose.orientation.x = q[0]
        self.odom_msgs_.pose.pose.orientation.y = q[1]
        self.odom_msgs_.pose.pose.orientation.z = q[2]
        self.odom_msgs_.pose.pose.orientation.w = q[3]
        self.odom_msgs_.pose.pose.position.x = self.x_
        self.odom_msgs_.pose.pose.position.y = self.y_
        self.odom_msgs_.header.stamp = self.get_clock().now().to_msg()
        self.odom_msgs_.twist.twist.linear.x= V
        self.odom_msgs_.twist.twist.angular.z = w

        self.odom_pub_.publish(self.odom_msgs_)

        self.transform_stamped_.transform.translation.x  = self.x_
        self.transform_stamped_.transform.translation.y = self.y_
        self.transform_stamped_.transform.rotation.x = q[0]
        self.transform_stamped_.transform.rotation.y = q[1]
        self.transform_stamped_.transform.rotation.z = q[2]
        self.transform_stamped_.transform.rotation.w = q[3]
        self.transform_stamped_.header.stamp =  self.get_clock().now().to_msg()
        self.br_.sendTransform(self.transform_stamped_)


        # self.get_logger().info("Linear Velociety measured by sensor: %f and angular velocity is %f " %(V,w))

        # self.get_logger().info(" x: %f , y: %f and theta %f " %(self.x_,self.y_,self.theta_))

def main():
    rclpy.init()

    noisy_controller = NoisyController()
    rclpy.spin(noisy_controller)
    
    noisy_controller.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()